flake.nix

```diff
@@ -3,62 +3,420 @@
+          # the expected version.
+          mkVersionCheck = { pkg, cmd }: runCommand "check-${pkg.name}-version"
+            {
+              PATH = "${pkg.out}";
+            } ''
+            actualVersion=$(${pkg.out}/bin/${cmd})
+            if [ "$actualVersion" != "${pkg.version}" ]; then
+              echo "expected ${pkg.name} version \"${pkg.version}\", got \"$actualVersion\""
+              exit 1
+            fi
+
+            # the check derivation must have an output.
+            touch $out
+          '';
+          # produces a check derivation that ensures a package's executable
+          # runs.
+          mkExecCheck = { pkg, cmd }: runCommand "check-${pkg.name}-${cmd}-exec"
+            { } ''
+            ${pkg.out}/bin/${cmd} && touch $out
+          '';
+        in
+        {
+          clickhouseVersion = mkVersionCheck
+            {
+              pkg = clickhouse;
+              cmd = "clickhouse server --version | cut -d ' ' -f 4";
             };
-        }
-      );
+
+          cockroachdbVersion = mkVersionCheck
+            {
+              pkg = cockroachdb;
+              cmd = "cockroach version --build-tag";
+            };
+
+          mgdCanExec = mkExecCheck {
+            pkg = mgd;
+            cmd = "mgd help";
+          };
+
+          dpdCanExec = mkExecCheck {
+            pkg = dendrite-stub;
+            cmd = "dpd help";
+          };
+
+          swadmCanExec = mkExecCheck {
+            pkg = dendrite-stub;
+            cmd = "swadm help";
+          };
+        };
+
+      devShells.x86_64-linux.default =
+        pkgs.mkShell.override
+          {
+            # use Clang as the C compiler for all C libraries
+            stdenv = pkgs.clangStdenv;
+          }
+          {
+            inherit buildInputs;
+            nativeBuildInputs = nativeBuildInputs ++ [
+              # Dendrite and maghemite, for running tests.
+              dendrite-stub
+              mgd
+              clickhouse
+              cockroachdb
+            ];
+
+            name = "omicron";
+            DEP_PQ_LIBDIRS = "${pkgs.postgresql.lib}/lib";
+            LIBCLANG_PATH = "${pkgs.libclang.lib}/lib";
+            OPENSSL_DIR = "${pkgs.openssl.dev}";
+            OPENSSL_LIB_DIR = "${pkgs.openssl.out}/lib";
+
+            MG_OPENAPI_PATH = mgOpenAPI;
+            DDM_OPENAPI_PATH = ddmOpenAPI;
+            DPD_OPENAPI_PATH = dendriteOpenAPI;
+
+            shellHook = ''
+              rm out/mgd
+              rm out/dendrite-stub
+              rm -r out/clickhouse
+              rm -r out/cockroachdb
+
+              mkdir -p out/clickhouse
+              mkdir -p out/cockroachdb/
+
+              ln -s ${mgd.out} -T out/mgd
+              ln -s ${dendrite-stub.out} -T out/dendrite-stub
+              ln -s ${clickhouse.out}/bin/clickhouse out/clickhouse/clickhouse
+              ln -s ${clickhouse.out}/etc/config.xml out/clickhouse
+              ln -s ${cockroachdb.out}/bin out/cockroachdb/bin
```

Comment:
fwiw, we don't _strictly_ need to symlink the `clickhouse` and `cockroach` executables into `out/`, because `omicron-dev` happily runs them from the `$PATH`; it's just Dendrite and Maghemite that need to be in `out/`, because Dendrite apparently has other stuff it needs in its filesysten,

but, i thought it was nicer to symlink them as well, to more closely mimic the behavior of the various `ci_download_whatever` scripts.

ideally, we could also just point everything at Dendrite and Maghemite's `nix-store` paths via env variables and not have to do any symlinking, but that would require changing other code that i'm not deeply familiar with yet. i might look into that next.